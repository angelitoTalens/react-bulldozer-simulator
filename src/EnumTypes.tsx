export enum Direction {
    Up = "U",
    Down = "D",
    Left = "L",
    Right = "R",
}

export enum Commands {
    RotateLeft = "L",
    RotateRight = "R",
    Advance = "A"
}

export type Position = {
    X: number, 
    Y: number
}

export enum LandType {
    Plain = "o",
    Rocky = "r",
    TreePlanted = "t",
    Protected = "T",
    Cleared = ""
}

export enum SimulationStatus {
    Ongoing,
    NoGrid,
    OutOfBounds,
    ClearedProtected,
    UserTerminated
}

export enum TabKeys {
    Simulation = "Simulation",
    Results = "Results"
}

export type Cost = {
    quantity: number;
    value: number;
}

export type LandVisitSummary = {
    clearPlainLand: number;
    revisitClearedLand: number;
    clearRockyLand: number;
    clearTreePlantedLand: number;
}

export type Movements = {
    command: Commands;
    steps: number;
}