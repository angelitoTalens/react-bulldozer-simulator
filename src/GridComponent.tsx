import React from "react";
import "./styles/GridComponent.scss"

import BlockComponent from "./BlockComponent"
import { LandType, Position, Direction } from "./EnumTypes"

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Col";

type GridComponentProps = {
    grid: LandType[][];
    position: Position;
    direction: Direction;
}

class GridComponent extends React.Component<GridComponentProps> {

    renderBlock(t:LandType, c:boolean, d:Direction, key:string) {
        return (<BlockComponent landType={t} currentPos={c} direction={d} key={key}/>);
    }

    renderGridMap() {
        let gridMap = [];
        let y = 0;
        for (let row of this.props.grid) {
            let rowBlocks = [];
            let x = 0;
            for (const block of row) {
                const isCurrent = ((this.props.position.X === x) && ((this.props.position.Y === y)));
                const key = y + "," + x;
                rowBlocks.push(this.renderBlock(block, isCurrent, this.props.direction, key));
                x++;
            }
            gridMap.push(<Row className="grid-row" key={y}>{rowBlocks}</Row>);
            y++;
        }

        return(
            <div>
                {gridMap}
            </div>
        );
    }
      

    render() {
        return (
            <Container>
                {this.renderGridMap()}
            </Container>
        );
    }
}

export default GridComponent;
