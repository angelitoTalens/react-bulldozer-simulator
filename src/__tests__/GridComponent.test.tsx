import React from 'react';
import renderer from 'react-test-renderer';

import GridComponent from '../GridComponent';
import { LandType, Direction } from '../EnumTypes';

test ("Test GridComponent", ()  => {

    const grid: LandType[][] = [
        [LandType.Plain, LandType.Rocky, LandType.TreePlanted],
        [LandType.Plain, LandType.Rocky, LandType.Protected],
        [LandType.Plain, LandType.Rocky, LandType.Cleared]
    ];

    const component = renderer.create(

        <GridComponent
            grid={grid}
            position={{X:1, Y:1}}
            direction={Direction.Right}
        />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});