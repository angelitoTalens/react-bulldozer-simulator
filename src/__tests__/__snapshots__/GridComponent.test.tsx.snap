// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GridComponent 1`] = `
<div
  className="container"
>
  <div>
    <div
      className="grid-row col"
    >
      <div
        className="block col"
      >
        o
      </div>
      <div
        className="block col"
      >
        r
      </div>
      <div
        className="block col"
      >
        t
      </div>
    </div>
    <div
      className="grid-row col"
    >
      <div
        className="block col"
      >
        o
      </div>
      <div
        className="block current-block col"
      >
        <img
          alt="bulldozer"
          src="bulldozer.gif"
          style={
            Object {
              "transform": "rotateY(180deg)",
            }
          }
        />
      </div>
      <div
        className="block col"
      >
        T
      </div>
    </div>
    <div
      className="grid-row col"
    >
      <div
        className="block col"
      >
        o
      </div>
      <div
        className="block col"
      >
        r
      </div>
      <div
        className="block col"
      >
        
      </div>
    </div>
  </div>
</div>
`;
